Translation of lifetimes in Fn trait input parameters

I think this is a fairly niche and easy to work around case but it's a surprising footgun that might be worth a fix?

```rust
use async_trait::async_trait;
use core::future::Future;

#[async_trait]
pub trait ProcessPool: Send + Sync {
    type ThreadPool;

    async fn spawn<F, Fut, T>(&self, work: F) -> T
    where
        F: FnOnce(&Self::ThreadPool) -> Fut + Send,
        Fut: Future<Output = T> + 'static;
}

#[async_trait]
impl<P: ?Sized> ProcessPool for &P
where
    P: ProcessPool,
{
    type ThreadPool = P::ThreadPool;

    async fn spawn<F, Fut, T>(&self, work: F) -> T
    where
        F: FnOnce(&Self::ThreadPool) -> Fut + Send,
        Fut: Future<Output = T> + 'static,
    {
        (*self).spawn(work).await
    }
}
```
[Playground](https://play.rust-lang.org/?edition=2018&gist=502159c0378036ce4f089079480c43de)

`Self::ThreadPool` on line 23 is translated to `<&P as ProcessPool>::ThreadPool`. I believe due to how unnamed lifetimes are treated in the call syntax of Fn traits, this is desugared to include a HRTB? This leads to the failure:

```
expected a `std::ops::FnOnce<(&<&P as ProcessPool>::ThreadPool,)>` closure, found `F`
```

This can be resolved by translating instead to `<&'static P as ProcessPool>::ThreadPool`. I don't know if this breaks other use cases however?

> This can be resolved by translating instead to `<&'static P as ProcessPool>::ThreadPool`. I don't know if this breaks other use cases however?

IIUC this breaks use cases like [`FutureExt::inspect`](https://docs.rs/futures/0.3.5/futures/future/trait.FutureExt.html#method.inspect):

```rust
use async_trait::async_trait;
use std::future::Future;

#[async_trait]
pub trait FutureExt: Future + Send + Sync {
    async fn inspect<F>(self, f: F) -> Self::Output
    where
        F: FnOnce(&Self::Output) + Send,
        Self: Sized;
}

#[async_trait]
impl<Fut: ?Sized + Future + Send + Sync> FutureExt for Fut {
    async fn inspect<F>(self, f: F) -> Fut::Output
    where
        F: FnOnce(&Fut::Output) + Send,
        Self: Sized,
    {
        let output = self.await;
        f(&output);
        output
    }
}
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f8679b7749e05e6abd73d1387e9062d1)

if using `&'static`:

```diff
-         F: FnOnce(&Fut::Output) + Send,
+         F: FnOnce(&'static Fut::Output) + Send,
```

```text
error[E0310]: the associated type `<Fut as std::future::Future>::Output` may not live long enough
   --> tests/test.rs:924:12
    |
924 |         F: FnOnce(&'static Fut::Output) + Send,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7dc36baca9c7a3753a47118753cb2588)

--- a/src/expand.rs
+++ b/src/expand.rs
@@ -81,6 +81,13 @@ pub fn expand(input: &mut Item, is_local: bool) {
             }
         }
         Item::Impl(input) => {
+            let mut lifetimes = CollectLifetimes::with("'impl_life");
+            lifetimes.visit_type_mut(&mut *input.self_ty);
+            lifetimes.visit_path_mut(&mut input.trait_.as_mut().unwrap().1);
+            let params = &input.generics.params;
+            let elided = lifetimes.elided;
+            input.generics.params = parse_quote!( #(#elided,)* #params );
+
             let context = Context::Impl {
                 impl_generics: &input.generics,
                 receiver: &input.self_ty,

--- a/src/lifetime.rs
+++ b/src/lifetime.rs
@@ -24,6 +24,7 @@ impl VisitMut for HasAsyncLifetime {
 pub struct CollectLifetimes {
     pub elided: Vec<Lifetime>,
     pub explicit: Vec<Lifetime>,
+    pub name: &'static str,
 }
 
 impl CollectLifetimes {
@@ -31,6 +32,15 @@ impl CollectLifetimes {
         CollectLifetimes {
             elided: Vec::new(),
             explicit: Vec::new(),
+            name: "'life",
+        }
+    }
+
+    pub fn with(name: &'static str) -> Self {
+        CollectLifetimes {
+            elided: Vec::new(),
+            explicit: Vec::new(),
+            name,
         }
     }
 
@@ -50,7 +60,7 @@ impl CollectLifetimes {
     }
 
     fn next_lifetime(&mut self) -> Lifetime {
-        let name = format!("'life{}", self.elided.len());
+        let name = format!("{}{}", self.name, self.elided.len());
         let life = Lifetime::new(&name, Span::call_site());
         self.elided.push(life.clone());
         life

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -905,3 +905,35 @@ mod issue104 {
 
     impl_t1!(Foo, 1);
 }
+
+// https://github.com/dtolnay/async-trait/issues/106
+mod issue106 {
+    use async_trait::async_trait;
+    use std::future::Future;
+
+    #[async_trait]
+    pub trait ProcessPool: Send + Sync {
+        type ThreadPool;
+
+        async fn spawn<F, Fut, T>(&self, work: F) -> T
+        where
+            F: FnOnce(&Self::ThreadPool) -> Fut + Send,
+            Fut: Future<Output = T> + 'static;
+    }
+
+    #[async_trait]
+    impl<P: ?Sized> ProcessPool for &P
+    where
+        P: ProcessPool,
+    {
+        type ThreadPool = P::ThreadPool;
+
+        async fn spawn<F, Fut, T>(&self, work: F) -> T
+        where
+            F: FnOnce(&Self::ThreadPool) -> Fut + Send,
+            Fut: Future<Output = T> + 'static,
+        {
+            (*self).spawn(work).await
+        }
+    }
+}

